### PR TITLE
Port the software to run the cylc workflow on derecho using cylc8.x.

### PR DESCRIPTION
**Pull request by jim-p-w**
### Description
These changes allow running the cylc workflow on derecho (and cheyenne).
This is based on work by Tao Sun in the Derecho-develop branch of this repo.

This change requires an anaconda environment with cylc8.x software on derecho. The `CYLC_ENV` environment variable may be set to the name of the conda environment which has cylc8.x software.  If CYLC_ENV is not set the default conda package /glade/work/jwittig/conda-envs/my-cylc8.2 will be used. 

Cylc 8.2 was tested.

Running on derecho will require setting the 'mpas bundle' in Build.py to point to software built on derecho. By default it points to a build in my workspace, which works for all of the predefined tests.

The priorities of jobs (as specified by the Queue, e.g. `CriticalQueue: economy`, `CriticalQueue: premium`, `NonCriticalQueue: premium`) will be honored on derecho, as will submitting small jobs to Casper.

The changes are grouped logically by commit, so it may be easier to review each commit (instead of just looking at all of the changed files at once).

### Issue addressed
Most of the issues in #(https://github.com/NCAR/MPAS-Workflow/issues/263) have been addressed, however the cylc configuration files are not yet cylc8 compliant. The changes for complying with the cylc 8 syntax will be in a separate PR.

#### Tier 1:  All tests were run on cheyenne and derecho.
 - [x] 3dvar_OIE120km_WarmStart
 - [x] 3denvar_OIE120km_IAU_WarmStart
 - [x] 3dvar_OIE120km_ColdStart
 - [x] 3dvar_O30kmIE60km_ColdStart
 - [x] 3denvar_O30kmIE60km_WarmStart
 - [x] eda_OIE120km_WarmStart
 - [x] getkf_OIE120km_WarmStart
 - [x] ForecastFromGFSAnalysesMPT

#### Tier 2 (optional):
 - [ ] GenerateGFSAnalyses
 - [ ] GenerateObs

### CODE REVIEW

**Comment by ibanos90** [comment]
Thanks for this PR @jim-p-w. Should we change the name of the environment files to `machine.sh/machine.csh` instead of `cheyenne.sh/cheyenne.csh`?

**Comment by jim-p-w** [comment]
> Thanks for this PR @jim-p-w. Should we change the name of the environment files to `machine.sh/machine.csh` instead of `cheyenne.sh/cheyenne.csh`?

Good idea, since the setup scripts are not machine specific.
Change is pushed.

File: env-setup/cheyenne.sh
Comment (RGNystrom) [inline env-setup/cheyenne.sh:1]: @jim-p-w 

What about setting your conda environment as the default by replacing the following?
`if [[ "$CYLC_ENV" == "" ]]; then`
`echo "set CYLC_ENV environment variable to the name of the cylc8 conda package"`
`return`
`fi`
with
`if [[ "$CYLC_ENV" == "" ]]; then`
`echo "CYLC_ENV not set"`
 `echo "setting CYLC_ENV = /glade/work/jwittig/conda-envs/my-cylc8.2"`
 `export CYLC_ENV="/glade/work/jwittig/conda-envs/my-cylc8.2"`
 `fi`

**Comment by mos3r3n** [comment]
In submit.csh (Line 51-56), to determine whether a suite is running, we can do as follows: 
```
set status = `cylc scan -t rich | grep -c "${SuiteName} "`
if ( $status > 0 ) then
  echo "$0 (INFO): a cylc suite named $SuiteName is already running!"
  echo "$0 (INFO): stopping the suite (30 sec.), then starting a new one..."
  cylc stop --kill MPAS-Workflow/$SuiteName
  sleep 30
else
  echo "$0 (INFO): confirmed that a cylc suite named $SuiteName is not already running"
  echo "$0 (INFO): starting a new suite..."
endif
```

JimW: Thanks Tao, change is pushed.

File: env-setup/cheyenne.sh
Comment (jim-p-w) [reply env-setup/cheyenne.sh:1]: Good idea! I'll do it.

File: config/environmentJEDI.csh [hunk 1]
```diff
@@ -1,37 +1,70 @@
 #!/bin/csh -f
 
 if ( $?config_environmentJEDI ) exit 0
+
+echo "Loading Spack-Stack 1.3.1"
```
Comment (RGNystrom) [inline config/environmentJEDI.csh]: Currently using spack-stack-1.5.1, should be updated

File: initialize/config/Config.py [hunk 2]
```diff
@@ -64,8 +66,8 @@ def get(self, dotSeparatedKey, typ=None, options=None):
     if options is not None and v is not None:
       assert isinstance(options, Iterable) and not isinstance(options, str), (
         'options must be a list of valid values or not present at all, not '+str(options))
-      assert v in options, ('invalid value for '+dotSeparatedKey+': '+str(v)+
-        '; choose one of '+str(options))
+      #assert v in options, ('invalid value for '+dotSeparatedKey+': '+str(v)+
+        #'; choose one of '+str(options))
```
Comment (RGNystrom) [inline initialize/config/Config.py]: I think these two lines that are commented out can just be removed.

File: initialize/framework/Build.py [hunk 3]
```diff
@@ -34,6 +35,43 @@ class Build(Component):
...
+
+      # Ungrib
+      wpsBuildDir = '/glade/work/jwittig/repos1/WPS/'
+      # Obs2IODA-v2
+      obs2iodaBuildDir = '/glade/work/taosun/Derecho/MPAS/Obs2IODA'
+      # Mean state calculator
+      meanStateBuildDir = self.variablesWithDefaults['mpas bundle'][0]+'/bin'
+    elif system == 'cheyenne':
+      self.variablesWithDefaults['mpas bundle'] = \
+        ['/glade/p/mmm/parc/liuz/pandac_common/mpas-bundle-code-build/mpas_bundle_2.0_gnuSP/build', str]
```
Comment (RGNystrom) [inline initialize/framework/Build.py:57]: This path with `/glade/p` and others below should be updated to campaign storage.

**Comment by RGNystrom** [comment]
All tests succeed for me, with the exception of `ObsToIoda` which is likely related to the build.

File: initialize/framework/Build.py [hunk 4]
```diff
@@ -34,6 +35,43 @@ class Build(Component):
...
+      self.variablesWithDefaults['mpas bundle'] = \
+          ['/glade/work/jwittig/repos1/mpas-bundle-r2.0/build-gnu-derecho-single/', str]
+      self.variablesWithDefaults['bundle compiler used'] = ['intel-cray', str,
+        ['gnu-cray', 'intel-cray']]
+      self.variablesWithDefaults['forecast directory'] = ['bundle', str]
+
+      # Ungrib
+      wpsBuildDir = '/glade/work/jwittig/repos1/WPS/'
+      # Obs2IODA-v2
+      obs2iodaBuildDir = '/glade/work/taosun/Derecho/MPAS/Obs2IODA'
```
Comment (ibanos90) [inline initialize/framework/Build.py]: `src` is missing in the path: `/glade/work/taosun/Derecho/MPAS/Obs2IODA/src`

Comment (mos3r3n) [reply initialize/framework/Build.py]: This executable is built with spack-stack gnu. I also made some modifications to let it not output data with bad QC flags. Be careful to use it.

File: env-setup/cheyenne.sh
Comment (jim-p-w) [reply env-setup/cheyenne.sh:1]: Changes are pushed.

File: config/environmentJEDI.csh (hunk 1 above)
Comment (jim-p-w) [reply config/environmentJEDI.csh]: Updated the output line, and use the same spack-stack version for both intel and gnu toolchains.
Change is pushed.

File: initialize/config/Config.py (hunk 2 above)
Comment (jim-p-w) [reply initialize/config/Config.py]: Done.

File: initialize/framework/Build.py (hunk 3 above)
Comment (jim-p-w) [reply initialize/framework/Build.py:57]: I left the references to /glade/p in the cheyenne specific code, as glade/p won't go away until cheyenne is retired.
No references to glade/p  remain in the derecho code.

File: initialize/framework/Build.py (hunk 4 above)
Comment (jim-p-w) [reply initialize/framework/Build.py]: I changed the path to point to your project space in /glade/csampaign.

File: initialize/framework/Experiment.py [hunk 5]
```diff
@@ -103,10 +103,11 @@ def __init__(self,
     self._msg('Setting up a new suite')
     self._msg('  SuiteName: '+self['SuiteName'])
     self._msg('  mainScriptDir: '+self['mainScriptDir'])
+    self._msg('  ExperimentDirectory: '+self['ExperimentDirectory'])
     self._msg('======================================================================')
     self._msg('')
 
-    cmd = ['rm', '-rf', self['mainScriptDir']]
+    cmd = ['rm', '-rf', self['ExperimentDirectory']]
```
Comment (jim-p-w) [inline initialize/framework/Experiment.py:110]: This change was made to remove the various artifacts created in the directories which are at the same level as MPAS-Workflow, e.g. in the Observations directory there are links such as 
```
  7 18:28 obs2ioda-v2.x -> /glade/campaign/mmm/parc/ivette/pandac/fork_obs2ioda/obs2ioda/obs2ioda-v2/src/obs2ioda-v2.x
```
which weren't getting cleaned up.
That's why some tests passed for me which failed for others.

File: env-setup/machine.sh [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+type conda >& /dev/null
+if [ $? -eq 0 ]; then
+  conda deactivate
+fi
+
+if [[ "$NCAR_HOST" == "derecho" ]]; then
+  if [[ "$CYLC_ENV" == "" ]]; then
+    echo 'CYLC_ENV environment variable is not set, setting it to /glade/work/jwittig/conda-envs/my-cylc8.2'
+    export CYLC_ENV=/glade/work/jwittig/conda-envs/my-cylc8.2
+    return
```
Comment (ibanos90) [inline env-setup/machine.sh]: Need to remove this `return`, otherwise the modules are not loaded

Comment (jim-p-w) [reply env-setup/machine.sh]: Indeed! Thanks for catching this.

**Comment by ibanos90** [comment]
Hi @jim-p-w, can you update the PR's description given that now we export CYCL_ENV inside the environment file? Other than that I think is ready to merge

File: env-setup/machine.sh (hunk 6 above)
Comment (jim-p-w) [reply env-setup/machine.sh]: Done.

**Review by ibanos90 — APPROVED** [review]
Thanks @jim-p-w for your work on this! All tests completed successfully.

**Review by RGNystrom — APPROVED** [review]
Looks good to me now, thanks @jim-p-w.

**Pull request merged**